### PR TITLE
Optimize markers calculations

### LINKED ISSUE #824: Map gets super laggy when there are a lot of Markers
Hi there!

I am currently making an app which will have *a lot of* markers - around 4000 spread across whole country. After some testing, it got obvious that all of this gets super laggy

When I use all 3.5k markers and zoom on area with *no markers at all*, and start dragging the map, I get ~12 fps, on *phisical Android tablet in `profile` mode*

### After some debugging, `print()`ing and `Stopwatch()`ing, I traced which parts use the most time:

In `lib/src/layer/marker_layer.dart` -> `MarkerLayer` -> `build()`, there is a `for-loop` that iterates over all given `Markerks`:
https://github.com/fleaflet/flutter_map/blob/f09b193505ad31eb496b99a379346d99ec944cc7/lib/src/layer/marker_layer.dart#L130

#### The whole thing was taking ~25ms, and with `Stopwatch()`, I traced which parts of loop use the most time:
```dart
// I did it like this:
int sumStep = 0;
for(var i in list){
  var w = Stopwatch()..start();
  i.someStep();
  sumStep += w.elapsedMicroseconds();
  w.reset();
  // Of course, some precision is lost, because some steps take 1.something microsecond
}
print('Step: $sumStep');  // Summed microseconds
```
This was taking around 7500 us:
https://github.com/fleaflet/flutter_map/blob/f09b193505ad31eb496b99a379346d99ec944cc7/lib/src/layer/marker_layer.dart#L131

This 3600:
https://github.com/fleaflet/flutter_map/blob/f09b193505ad31eb496b99a379346d99ec944cc7/lib/src/layer/marker_layer.dart#L132-L133

And the rest around 0:
https://github.com/fleaflet/flutter_map/blob/f09b193505ad31eb496b99a379346d99ec944cc7/lib/src/layer/marker_layer.dart#L140-L152

### So I started to dig into `.project()`:
https://github.com/fleaflet/flutter_map/blob/f09b193505ad31eb496b99a379346d99ec944cc7/lib/src/map/map.dart#L417-L420

https://github.com/fleaflet/flutter_map/blob/f09b193505ad31eb496b99a379346d99ec944cc7/lib/src/geo/crs/crs.dart#L26-L34

`this.scale()` and `transformation.transform()` don't seem to be expensive - they are only some basic calculations...

So, `projection.project`:
https://github.com/fleaflet/flutter_map/blob/f09b193505ad31eb496b99a379346d99ec944cc7/lib/src/geo/crs/crs.dart#L450-L456

https://github.com/maRci002/proj4dart/blob/7eb11da840ce21c78c017e0a89ee0332ba5a60e3/lib/src/classes/projection.dart#L152

Okay... this ^ is *big*, and not even a part of this repo, so I don't think we can do anything about it...

# Conclusions
While re-building `MarkerLayer` (which happens constantly while dragging the map), we are doing some expensive calculations, which slow down the whole map - that is, calculating some `position` from `LatLng`

// I didn't dig into `pos.multiply` (step which was taking 3600us), and I don't really know what it's doing...

I'm fresh into this project and I don't know much about mapping, but it seems like we could somehow cache just the `projection`, and apply only `scale` and `transformations` - since only them *really change* when you drag the map

(This generally seems related #768, but discussion there got kinda chaotic, not knowing what issue actually was, so I'm making this separate)

### PR DESCRIPTION
**Pull request by TheLastGimbus**
This PR is attempting to optimize and shrink all calculations that are done when user drags map with **a lot of** markers

Hopefully, I will get it to state where we have full 60 fps with 5k markers, but we will se how it goes

Fixes #824 ~~I would really want to~~ Yeah, pretty much - I don't see we how could do much better than this

### CODE REVIEW

**Comment by TheLastGimbus** [comment]
Travis CI is broken...

**Comment by TheLastGimbus** [comment]
You literally commented that when I was about to push it :smile:

File: lib/src/layer/marker_layer.dart [hunk 1]
```diff
@@ -110,44 +110,34 @@ class MarkerLayer extends StatelessWidget {
...
+          final height = marker.height - marker.anchor.top;
+          var sw = CustomPoint(pxPoint.x + width, pxPoint.y - height);
+          var ne = CustomPoint(pxPoint.x - width, pxPoint.y + height);
 
-          if (!_boundsContainsMarker(markerOpt)) {
+          if (!map.pixelBounds.containsPartialBounds(Bounds(sw, ne))) {
             continue;
           }
 
+          final pos = pxPoint.multiplyBy(map.getZoomScale(map.zoom, map.zoom)) -
```
Comment (TheLastGimbus) [inline lib/src/layer/marker_layer.dart:131]: Like... what the hell is that? `map.getZoomScale(map.zoom, map.zoom)` will always return a `1.0`

# :thinking:

**Review by TheLastGimbus — COMMENTED** [review]
This may look kinda hack-ish, but this actually brough whole loop's time (on my tablet) from ~20 to ~10ms! On 5k markers!

(As always: When zoomed to no markers and not zooming in/out)

**Comment by TheLastGimbus** [comment]
@ibrierley You proposed to cache them per-zoom-level. Out of curiosity, I printed when zoom changes and:
```
I/flutter (16890): 6.249008887924742
I/flutter (16890): 6.254358513273377
I/flutter (16890): 6.255482934237276
I/flutter (16890): 6.261930705947585
I/flutter (16890): 6.266114410749081
I/flutter (16890): 6.2766681718506305
I/flutter (16890): 6.283982456742614
I/flutter (16890): 6.2892041742113065
I/flutter (16890): 6.296456503011495
I/flutter (16890): 6.301633471546501
I/flutter (16890): 6.303673583816311
I/flutter (16890): 6.30882487202124
I/flutter (16890): 6.311932043194194
I/flutter (16890): 6.317053908214326
I/flutter (16890): 6.3190726270623445
```
We probably couldn't phisically fit so many `int`s into whole device ram :laughing: So I guess that's the best we can settle on

**Comment by TheLastGimbus** [comment]
Okay, this is ready for review and merge

> Hopefully, I will get it to state where we have full 60 fps with 5k 

I have 30 at 5k and 60 at 2.5k - not bad :+1:

**Comment by TheLastGimbus** [comment]
### One important thing I noticed:

When you have a simple widget as markers, and make it a `const`, `fps`es can sometimes *double* when you have all of them on the screen

We could add this hint somewhere in the documentation

(but, of course, behavior between devices/widget cases can vary)

File: lib/src/layer/marker_layer.dart [hunk 2]
```diff
@@ -107,18 +107,16 @@ class MarkerLayer extends StatelessWidget {
...
 
-  MarkerLayer(this.markerOpts, this.map, this.stream)
-      : super(key: markerOpts.key);
-
-  bool _boundsContainsMarker(Marker marker) {
-    var pixelPoint = map.project(marker.point);
-
-    final width = marker.width - marker.anchor.left;
-    final height = marker.height - marker.anchor.top;
+  // Note: I don't know if it's okay to store mutable stuff like this in stless
```
Comment (johnpryan) [inline lib/src/layer/marker_layer.dart:110]: It's not; this will be discarded anytime this widget is rebuilt.

Comment (johnpryan) [reply lib/src/layer/marker_layer.dart:110]: Convert this into a StatefulWidget and clear the pxCache and lastZoom when the `didUpdateWidget` lifecycle method is called.

File: lib/src/layer/marker_layer.dart [hunk 3]
```diff
@@ -127,30 +125,39 @@ class MarkerLayer extends StatelessWidget {
       stream: stream, // a Stream<int> or null
       builder: (BuildContext context, AsyncSnapshot<int> snapshot) {
         var markers = <Widget>[];
-        for (var markerOpt in markerOpts.markers) {
-          var pos = map.project(markerOpt.point);
-          pos = pos.multiplyBy(map.getZoomScale(map.zoom, map.zoom)) -
-              map.getPixelOrigin();
+        final sameZoom = map.zoom == lastZoom;
+        var i = -1;
+        for (var marker in markerOpts.markers) {
```
Comment (johnpryan) [inline lib/src/layer/marker_layer.dart:130]: Change to `for var i = 0; i < markerOpts.markers.length; i++)`

**Review by johnpryan — CHANGES_REQUESTED** [review]

File: lib/src/layer/marker_layer.dart (hunk 1 above)
Comment (kengu) [reply lib/src/layer/marker_layer.dart]: I agree, it looks a bit strange and unesserary @johnpryan? I see the same in all layer types, not just in marker_layer.
image
Only real use is when layer and map is at different zoom levels, as is the case in tile_layer.

**Comment by KlausJokisuo** [comment]
Any news on this?

**Comment by TheLastGimbus** [comment]
I was just about to apologize that I'm running late with pushing commits as always, but, oh, it's not me :smile: 

We are just waiting for @johnpryan review :+1:

**Comment by KlausJokisuo** [comment]
> I was just about to apologize that I'm running late with pushing commits as always, but, oh, it's not me 😄
> 
> We are just waiting for @johnpryan review 👍

No worries mate, what kind of performance improvements we can except for example with 2000 markers and 200 visible.

**Comment by TheLastGimbus** [comment]
Yeah, not that much, but it's always something

I think we need to make some new plugin for markers that will base on `Canvas` or something for *that many* markers

**Comment by KlausJokisuo** [comment]
> Yeah, not that much, but it's always something
> 
> I think we need to make some new plugin for markers that will base on `Canvas` or something for _that many_ markers
> 
> Edit: Oh, if that was a question - on my tablet, 3,5k markers **where** ~20fps when zoomed to 30 visible, and like 15fps when 200 visible. After optimization, 3.5k 60fps when <100 visible, then maybe 45fps when 200 visible 👍

Thanks, sorry I phrased my question oddly. 😅

File: lib/src/layer/marker_layer.dart (hunk 2 above)
Comment (kengu) [reply lib/src/layer/marker_layer.dart]: @TheLastGimbus Did you the the request for change from @johnpryan ?

**Review by kengu — APPROVED** [review]
All RFCs from @johnpryan are resolved in code. LGTM.

**Comment by kengu** [comment]
@TheLastGimbus The only remaining thing to do is to resolve the conflicts with master in lib/src/layer/marker_layer.dart

**Comment by TheLastGimbus** [comment]
Done! All working, markers rotating correctly :ok_hand: @kengu ready to merge :+1:

**Comment by kengu** [comment]
@TheLastGimbus I got this error in the log when opening the "A lot of markers" page in the example-app. 

```
======== Exception caught by widgets library =======================================================
The following RangeError was thrown building ManyMarkersPage(dirty, state: _ManyMarkersPageState#bb88c):
RangeError (end): Invalid value: Only valid value is 0: 500

The relevant error-causing widget was: 
  ManyMarkersPage file:///Users/kengu/src/git/flutter_map/example/lib/main.dart:66:45
When the exception was thrown, this was the stack: 
#0      RangeError.checkValidRange (dart:core/errors.dart:333:9)
#1      List.sublist (dart:core-patch/growable_array.dart:84:38)
#2      _ManyMarkersPageState.build (package:flutter_map_example/pages/many_markers.dart:85:56)
#3      StatefulElement.build (package:flutter/src/widgets/framework.dart:4612:27)
#4      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4495:15)
```

This fixes it
```
allMarkers.sublist(0, min(allMarkers.length, _sliderVal))
```

@johnpryan I've testet this on my phone and it looks good. The merge is blocked by your request for change, which I think is mostly resolved.

**Comment by TheLastGimbus** [comment]
# Done :100:

**Comment by TheLastGimbus** [comment]
@kengu you can pretty much press merge :ok_hand: 
<sub>I think</sub>

**Comment by kengu** [comment]
> @kengu you can pretty much press merge 👌
> I think

Waiting on @johnpryan - unable to merge until changes are ack-ed by him

**Review by johnpryan — APPROVED** [review]

**Pull request merged**